text/0795-ember-data-return-promise-save.md

```diff
@@ -0,0 +1,70 @@
+
+## Detailed design
+
+Introduce a new feature flag, `DS_MODEL_SAVE_PROMISE`, Model.save() will return an an Promise wrapper that resolves with the model if the save was successful, or rejects with an error if the save fails. Moreover, `isPending`, `isResolved`, and `isRejected` will be exposed as derived state on the promise.  When disabled, a PromiseObject will be returned to keep today's behavior, but accessing properties on the PromiseObject or calling non-Promise functions (.then, .finally, .catch) on the PromiseObject will issue a deprecation warning.
+
+Deprecation Plan:
+
+The deprecations can be implemented by adding deprecations around the existing PromiseObject. However, we can't include the deprecations in the actual PromiseObject class because [it is the base class for the relationship proxies](https://github.com/emberjs/data/blob/master/addon/-legacy-private/system/promise-proxies.js#L85). Instead, we can create and use a new DeprecatedPromiseObject for the DS.Model.save() function. It functions like a PromiseObject but deprecates the following behavior:
+
+    * Accessing properties on the PromiseObject returned from DS.Model.save()
+        * Use deprecateProperty from Ember to deprecate all the properties that the PromiseProxy Mixin provides. The list of properties can be found here: https://github.com/emberjs/ember.js/blob/master/packages/ember-runtime/lib/mixins/promise_proxy.js
+        * Deprecate accessing any unknown properties. For instance, someone might try to grab the value of an attr or computed property from a model, e.g. proxy.get('email').
+    * Calling functions that are not available on Promises
+        * Not Safe (issue deprecation): .get/.set/anything else really
+        * Safe (do not issue a deprecation): .then, .catch, .finally
+
+The implementation follows the standard feature-flagging process in Ember and Ember Data. The final implementation will look something like:
+
+```
+  save(options) {
+    const savePromise = this._internalModel.save(options).then(() => this);
+    
+    if (DS_MODEL_SAVE_PROMISE) {
+      return savePromise;
+    } else {
+      return DeprecatedPromiseObject.create({
+        promise: savePromise
+      });
+    }
+  },
+```
+
+## How we teach this
+
+We do not believe this requires any update to the Ember curriculum. API documentation does not need to be updated as it is already documented as returning a promise. This behavior of accessing properties through the PromiseProxyMixin specifically for DS.Model.save() is not documented in the guides.
+
+## Drawbacks
+
+For users relying on this behavior, they may have to refactor their code to access model properties and methods, especially if they were reaching for private APIs.
+
+## Alternatives
+
+The impact of not doing this prevents further changes in Ember Data.
```

Comment:
alternatives also include:

- returning just a promise
- returning a promise but also exposing a template/js helper in a new package that wraps any promise exposing these flags providing broader utility without imposing this promise-state management on ember-data itself.